dask/config.py

```diff
@@ -0,0 +1,194 @@
+        if k not in old and type(v) is dict:
+            old[k] = {}
+
+        if type(v) is dict:
+            update(old[k], v)
+        else:
+            old[k] = v
+    return old
+
+
+def merge(*dicts):
+    """ Update a sequence of nested dictionaries
+
+    This prefers the values in the latter dictionaries to those in the former
+
+    Example
+    -------
+    >>> a = {'x': 1, 'y': {'a': 2}}
+    >>> b = {'y': {'b': 3}}
+    >>> merge(a, b)
+    {'x': 1, 'y': {'a': 2, 'b': 3}}
+
+    See Also
+    --------
+    update
+    """
+    result = {}
+    for d in dicts:
+        update(result, d)
+    return result
+
+
+def collect_yaml(paths=config_paths):
+    """ Collect configuration from yaml files
+
+    This searches through ``config_paths``, expands to find all yaml or json
+    files, and then parses each file.
+    """
+    # Find all paths
+    file_paths = []
+    for path in paths:
+        if os.path.exists(path):
+            if os.path.isdir(path):
+                file_paths.extend(sorted([
+                    p for p in os.listdir(path)
+                    if os.splitext(p)[1].lower() in ('json', 'yaml', 'yml')
+                ]))
+            else:
+                file_paths.append(path)
+
+    configs = []
+
+    # Parse yaml files
+    for path in file_paths:
+        with open(path) as f:
+            data = yaml.load(f.read() or {})
+            configs.append(data)
+
+    return configs
+
+
+def collect_env():
+    """ Collect config from environment variables
+
+    This grabs environment variables of the form "DASK_FOO_BAR=123"
+    and turns these into config variables of the form ``{"foo-bar": 123}``
+    """
+    env = {}
+    for name, value in os.environ.items():
+        if name.startswith('DASK_'):
+            varname = name[5:].lower().replace('_', '-')
+            env[varname] = _parse_env_value(value)
+
+    return env
+
+
+def _parse_env_value(value):
+    """ Convert a string to an integer, float or boolean (in that order) if possible. """
+    bools = {
+        'true': True,
+        'false': False
+    }
+    try:
+        return int(value)
+    except ValueError:
+        pass
+    try:
+        return float(value)
+    except ValueError:
+        pass
+    return bools.get(value.lower(), value)
+
+
+def ensure_config_file(
+        source,
+        destination=os.path.join(os.path.expanduser('~'), '.config', 'dask')):
+    """ Copy file to default location if it does not already exist
+
+    This tries to move a default configuration files to a default location if
+    if does not already exist.
+
+    This is to be used by downstream modules (like dask.distributed) that may
+    have default configuration files that they wish to include in the default
+    configuration path.
+    """
+    if not os.path.exists(destination):
+        import shutil
+        if not os.path.exists(os.path.dirname(destination)):
+            try:
+                os.mkdir(os.path.dirname(destination))
```

Comment:
Maybe prettier to use our own `makedirs(path, exists_ok=True)`, since FileExistsError is a port anyway in py2.